test/core/tsi/ssl_transport_security_util_test.cc

```diff
@@ -0,0 +1,353 @@
+
+  EVP_PKEY_free(server_key);
+  BIO_free(server_key_bio);
+  X509_free(server_cert);
+  BIO_free(server_cert_bio);
+
+  // Configure both client and server to request (and accept any)
+  // certificate but fail if none is sent.
+  SSL_CTX_set_verify(client_ctx,
+                     SSL_VERIFY_PEER | SSL_VERIFY_FAIL_IF_NO_PEER_CERT,
+                     /*callback=*/nullptr);
+  SSL_CTX_set_cert_verify_callback(client_ctx, VerifySucceed,
+                                   /*arg=*/nullptr);
+  SSL_CTX_set_verify(server_ctx,
+                     SSL_VERIFY_PEER | SSL_VERIFY_FAIL_IF_NO_PEER_CERT,
+                     /*callback=*/nullptr);
+  SSL_CTX_set_cert_verify_callback(server_ctx, VerifySucceed,
+                                   /*arg=*/nullptr);
+
+  // Turns off the session caching.
+  SSL_CTX_set_session_cache_mode(client_ctx, SSL_SESS_CACHE_OFF);
+  SSL_CTX_set_session_cache_mode(server_ctx, SSL_SESS_CACHE_OFF);
+
+  // Set both the min and max TLS version to 1.3
+  SSL_CTX_set_min_proto_version(client_ctx, TLS1_3_VERSION);
+  SSL_CTX_set_min_proto_version(server_ctx, TLS1_3_VERSION);
+  SSL_CTX_set_max_proto_version(client_ctx, TLS1_3_VERSION);
+  SSL_CTX_set_max_proto_version(server_ctx, TLS1_3_VERSION);
+
+  // Create client and server connection objects and configure their BIOs.
+  SSL* client(SSL_new(client_ctx));
+  SSL* server(SSL_new(server_ctx));
+
+  SSL_CTX_free(client_ctx);
+  SSL_CTX_free(server_ctx);
+
+  // Turns off issuance of session tickets by servers.
+  SSL_set_options(client, SSL_OP_NO_TICKET);
+  SSL_set_options(server, SSL_OP_NO_TICKET);
+
+  SSL_set_connect_state(client);
+  SSL_set_accept_state(server);
+  BIO* bio1;
+  BIO* bio2;
+  BIO_new_bio_pair(&bio1, /*writebuf1=*/0, &bio2, /*writebuf2=*/0);
+  SSL_set_bio(client, bio1, bio1);
+  SSL_set_bio(server, bio2, bio2);
+
+  // Drive both the client and server handshake operations to completion.
+  while (true) {
+    int client_ret = SSL_do_handshake(client);
+    int client_err = SSL_get_error(client, client_ret);
+    if (client_err != SSL_ERROR_NONE && client_err != SSL_ERROR_WANT_READ &&
+        client_err != SSL_ERROR_WANT_WRITE &&
+        client_err != SSL_ERROR_PENDING_TICKET) {
+      return absl::InternalError(
+          absl::StrCat("Client error:", SSL_error_description(client_err)));
+    }
+
+    int server_ret = SSL_do_handshake(server);
+    int server_err = SSL_get_error(server, server_ret);
+    if (server_err != SSL_ERROR_NONE && server_err != SSL_ERROR_WANT_READ &&
+        server_err != SSL_ERROR_WANT_WRITE &&
+        server_err != SSL_ERROR_PENDING_TICKET) {
+      return absl::InternalError(
+          absl::StrCat("Server error:", SSL_error_description(server_err)));
+    }
+    if (client_ret == 1 && server_ret == 1) {
+      break;
+    }
+  }
+
+  *out_client = client;
+  *out_server = server;
+
+  return absl::OkStatus();
+}
+
+std::size_t CalculateRecordSizeFromHeader(uint8_t fourth_header_byte,
```

Comment:
this can be a static function as well

Reply:
Need to achieve an agreement on helper function `SslFrameProtectorUtilFunctionTestHelper` before move this one.